### ISSUE TITLE
[curl] Find from brew, system, or vcpkg?

### ISSUE THREAD

**Issue by SamuelMarks** [microsoft/vcpkg#19669]
I've been using this for a while https://github.com/SamuelMarks/curl-simple-https/blob/90e16a9/cmake/FindCurlCustom.cmake

Also hacked together things like this before:
```cmake
if (APPLE)
  # brew
  if (IS_DIRECTORY "/usr/local/opt/curl")
    list(APPEND CMAKE_PREFIX_PATH  "/usr/local/opt/curl/bin")
    list(APPEND CMAKE_LIBRARY_PATH "/usr/local/opt/curl/lib")
  endif ()
endif ()
```

Now really it would be good to have whatever version is installed "just work", with a preference for whatever is specified in the toolchain file (if offered), then brew, then OS level.

One confusing thing is:
```cmake
find_package(CURL CONFIG REQUIRED)
target_link_libraries("${LIBRARY_NAME}" PRIVATE CURL::libcurl)
```

The whole `CURL::libcurl` bit. Unfortunately `find_package(CURL QUIET COMPONENTS libcurl CONFIG)` stopped working; I think from a recent vcpkg upgrade.

How is this meant to work?

**Comment by autoantwort** [comment]
So 
```cmake
find_package(CURL CONFIG REQUIRED)
target_link_libraries("${LIBRARY_NAME}" PRIVATE CURL::libcurl)
```
works? The supported COMPONENTS you can lookup the the cmake documentation: https://cmake.org/cmake/help/latest/module/FindCURL.html

**Comment by dg0yt** [comment]
Please note that there are two different modes of `find_package`.
- Module mode. Old school. This is what is provided by module `FindCURL` from CMake itself (i.e. depends on CMake version), and it needs to get complementary information from `vcpkg_cmake_wrapper.cmake` in vcpkg port curl. Documentation is part of CMake.
  ~~~cmake
  find_package(CURL MODULE ...)
  ~~~
  will look only for a find module.
- Config mode. Modern CMake. This what is provided by curl (upstream) itself (i.e. depends on curl version).
  ~~~cmake
  find_package(CURL CONFIG ...)
  ~~~
  will look only for config. Documentation should be part of curl. I don't know if it knows about "components".

**Comment by dg0yt** [comment]
> Unfortunately `find_package(CURL QUIET COMPONENTS libcurl CONFIG)` stopped working;

For more information, remove the `QUIET` and/or share logs. The config also depends on dependencies (port features).

**Comment by SamuelMarks** [comment]
Interesting, turning off `QUIET` gave me:

> CMake Warning at vcpkg/installed/x86-windows/share/curl/vcpkg-cmake-wrapper.cmake:7 (_find_package):
>   Found package configuration file:
> 
>     vcpkg/installed/x86-windows/share/curl/CURLConfig.cmake
> 
>   but it set CURL_FOUND to FALSE so package "CURL" is considered to be NOT
>   FOUND.

**Comment by dg0yt** [comment]
I guess this is caused by a missing transitive dependency. CMake should create its own log files. And probably there should be some output lines before the CURL line.

**Comment by NancyLi1013** [comment]
Hi @SamuelMarks 

Thanks for posting this issue.

Have you tried to use the following usage after you rebuild with the latest vcpkg?
```
find_package(CURL CONFIG REQUIRED)
target_link_libraries("${LIBRARY_NAME}" PRIVATE CURL::libcurl)
```

I also wonder why you need to add `COMPONENTS libcurl` in your project? Does the usage provided by vcpkg not meet your demand?

**Comment by SamuelMarks** [comment]
On the `COMPONENT libcurl` side… doesn't this prevent messing with the cache, so if my 3rd party dep is used in someones project, and they also depend on curl directly, as long as they are using a different component it won't conflict with the configuration options I've given to my dependency?

**Comment by dg0yt** [comment]
Hm, AFAICT using components was more likely to cause problems then to resolve them. It only works well with separate subdirectories, because there is only one `<Pkg>_FOUND` which is inherited to subsequent `add_directory`. I.e. once CMake found one component, it cannot look for another, in the same tree.

**Comment by NancyLi1013** [comment]
@SamuelMarks 

Since currently, `curl` uses CONFIG instead of MODULE mode to find `curl `in vcpkg. If you have the request for `COMPONENT libcurl` in CONFIG mode, please report this to upstream.

**Comment by NancyLi1013** [comment]
As you can from this, if adding `COMPONENT libcurl`, the error points to CURLConfig.cmake. This is upstream behavior.

```
CMake Error at F:/new/vcpkg/installed/x86-windows/share/curl/vcpkg-cmake-wrapper.cmake:7 (_find_package):
  Found package configuration file:

    F:/new/vcpkg/installed/x86-windows/share/curl/CURLConfig.cmake

  but it set CURL_FOUND to FALSE so package "CURL" is considered to be NOT
  FOUND.
Call Stack (most recent call first):
  F:/new/vcpkg/scripts/buildsystems/vcpkg.cmake:738 (include)
  CMakeLists.txt:8 (find_package)
```

**Comment by NancyLi1013** [comment]
Is this still a problem for your now? @SamuelMarks

**Comment by SamuelMarks** [comment]
Hmm, my script just broke on vcpkg. Let me try just the regular `find_package` now and see if it can find it from vcpkg and brew

**Comment by dg0yt** [comment]
Outside of vcpkg, `find_package(CURL CONFIG REQUIRED)` will only work if a curl config file is available. For curl, it should be safe to use
~~~
find_package(CURL REQUIRED)
~~~
which will also use the find module from CMake. In any case, you will have to take care that either libraries and headers, or config files, are in the suitable search paths.

**Comment by SamuelMarks** [comment]
@dg0yt Yeah that's what my hacky workaround one does:
https://github.com/offscale/libacquire/blob/master/cmake/FindCurlCustom.cmake

(but there has to be a better way)

**Comment by dg0yt** [comment]
Showing your code doesn't explain *why* you believe that a single `find_package(CURL REQUIRED)` is enough to find if curl is available or not. Which version of cmake, or which provider of curl, doesn't work in its prefered form of integration?

**Comment by SamuelMarks** [comment]
What I'm thinking is a find_package for curl that works, and that ordinally prioritises:

  0. toolchain provided (e.g., vcpkg)
  1. package manager provided (e.g., brew, apt)
  2. OS provided (common search paths)

Is that unreasonable to expect?

**Comment by dg0yt** [comment]
> What I'm thinking is a find_package for curl that works, and that ordinally prioritises:
> 
>     1. toolchain provided (e.g., vcpkg)
> 
>     2. package manager provided (e.g., brew, apt)
> 
>     3. OS provided (common search paths)
> 
> 
> Is that unreasonable to expect?

Not unreasonable from an abstract point of view. But in reality, you have to play by the rules of package search *as implemented* in `cmake`.  Don't ask me about about brew (`/usr/local`). But note that `apt` is managing the *OS provided* packages.

I still don't know "*why* you believe that a single `find_package(CURL REQUIRED)` is _not_ enough", i.e. the basic signature. 

If the basic signature doesn't work for a particular environment, you can modify it.
- For a particular package:
  - Set `<PackageName>_ROOT`
  - Set `<PackageName>_DIR`
  - Add either `CONFIG` or `NO_CONFIG`/`MODULE` to select a particular mode.
  - Add a `CONFIG` lookup before the basic lookup to change the default order.
- Globally:
  - Set the search path variables appropriately.
  - Set `CMAKE_FIND_PACKAGE_PREFER_CONFIG`

Note that setting of variables is a user option, i.e. as parameter to `cmake` configuration.
IMO, "set the search path variables appropriately" is the preferred option in cmake. And with the vcpkg toolchain, you don't even have to care about this.

**Comment by dg0yt** [comment]
One more remark:
CMake's `FindCURL.cmake` offers to ask for `COMPONENTS` but this is broken in vcpkg wrapper. Discovered while working on #22258. Fix cherry-picked to #22445.
However, there is no component `libcurl`. Cf. https://cmake.org/cmake/help/v3.21/module/FindCURL.html.

**Comment by Cheney-W** [comment]
Thanks for dg0yt's comment.
